gptme/server/tasks_api.py

```diff
@@ -0,0 +1,724 @@
+    metadata: dict[str, Any] = field(default_factory=dict)
+
+
+def get_tasks_dir() -> Path:
+    """Get the tasks storage directory."""
+    return get_logs_dir().parent / "tasks"
+
+
+def load_task(task_id: str) -> Task | None:
+    """Load a task from storage."""
+    tasks_dir = get_tasks_dir()
+    task_file = tasks_dir / f"{task_id}.json"
+
+    if not task_file.exists():
+        return None
+
+    try:
+        with open(task_file) as f:
+            data = json.load(f)
+        return Task(**data)
+    except Exception as e:
+        logger.error(f"Error loading task {task_id}: {e}")
+        return None
+
+
+def save_task(task: Task) -> None:
+    """Save a task to storage."""
+    tasks_dir = get_tasks_dir()
+    tasks_dir.mkdir(parents=True, exist_ok=True)
+
+    task_file = tasks_dir / f"{task.id}.json"
+
+    try:
+        with open(task_file, "w") as f:
+            json.dump(asdict(task), f, indent=2)
+    except Exception as e:
+        logger.error(f"Error saving task {task.id}: {e}")
+        raise
+
+
+def delete_task_file(task_id: str) -> bool:
+    """Delete a task file from storage."""
+    tasks_dir = get_tasks_dir()
+    task_file = tasks_dir / f"{task_id}.json"
+
+    if task_file.exists():
+        try:
+            task_file.unlink()
+            return True
+        except Exception as e:
+            logger.error(f"Error deleting task {task_id}: {e}")
+            return False
+    return False
+
+
+def list_tasks() -> list[Task]:
+    """List all tasks from storage."""
+    tasks_dir = get_tasks_dir()
+
+    if not tasks_dir.exists():
+        return []
+
+    tasks = []
+    for task_file in tasks_dir.glob("*.json"):
+        task_id = task_file.stem
+        task = load_task(task_id)
+        if task:
+            tasks.append(task)
+
+    return sorted(tasks, key=lambda t: t.created_at, reverse=True)
+
+
+def derive_task_status(task: Task) -> TaskStatus:
+    """Derive task status from conversation state and process information."""
+    if not task.conversation_ids:
+        return "pending"
+
+    try:
+        # Get the latest conversation
+        latest_conv_id = task.conversation_ids[-1]
+
+        # Check if conversation exists
+        logdir = get_logs_dir() / latest_conv_id
+        if not logdir.exists():
+            return "failed"
+
+        # Load conversation without lock for status check
+        manager = LogManager.load(latest_conv_id, lock=False)
+
+        # Check if process is active by looking for lock file
+        lock_file = logdir / ".lock"
+        if lock_file.exists():
+            try:
+                # Try to read the lock file to see if process is active
+                with open(lock_file):
+                    # If we can read it without blocking, no active process
+                    return "active"
+            except OSError:
+                # If we can't read it, likely an active process
+                return "active"
```

Comment:
Might be better to do this check on the client-side, since the conversation generating state is already exposed/subscribed to via the SSE stream.